File: voca_crm_api/src/main/java/com/vocacrm/api/service/StatisticsService.java
package com.vocacrm.api.service;

import com.vocacrm.api.dto.ChartDataDTO;
import com.vocacrm.api.dto.HomeStatisticsDTO;
import com.vocacrm.api.dto.RecentActivityDTO;
import com.vocacrm.api.dto.TodayScheduleDTO;
import com.vocacrm.api.model.BusinessPlace;
import com.vocacrm.api.model.Reservation;
import com.vocacrm.api.repository.BusinessPlaceRepository;
import com.vocacrm.api.repository.ReservationRepository;
import com.vocacrm.api.repository.StatisticsRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class StatisticsService {

    private final StatisticsRepository statisticsRepository;
    private final BusinessPlaceRepository businessPlaceRepository;
    private final ReservationRepository reservationRepository;
    private final JdbcTemplate jdbcTemplate;

    public HomeStatisticsDTO getHomeStatistics(String businessPlaceId) {
        BusinessPlace businessPlace = businessPlaceRepository.findById(businessPlaceId)
                .orElseThrow(() -> new RuntimeException("Business place not found"));

        Long todayReservations = reservationRepository.countTodayReservations(businessPlaceId);
        Integer todayVisits = statisticsRepository.getTodayVisitCount(businessPlaceId);
        Integer pendingMemos = statisticsRepository.getPendingMemosCount(businessPlaceId);
        Integer totalMembers = statisticsRepository.getTotalMembersCount(businessPlaceId);

        return HomeStatisticsDTO.builder()
                .businessPlaceId(businessPlaceId)
                .businessPlaceName(businessPlace.getName())
                .todayReservations(todayReservations != null ? todayReservations.intValue() : 0)
                .todayVisits(todayVisits != null ? todayVisits : 0)
                .pendingMemos(pendingMemos != null ? pendingMemos : 0)
                .totalMembers(totalMembers != null ? totalMembers : 0)
                .build();
    }

    public List<RecentActivityDTO> getRecentActivities(String businessPlaceId, Integer limit) {
        if (limit == null || limit <= 0) {
            limit = 10;
        }

        String sql = "SELECT * FROM get_recent_activities(?, ?)";

        return jdbcTemplate.query(sql,
                (rs, rowNum) -> RecentActivityDTO.builder()
                        .activityId(rs.getString("activity_id"))
                        .activityType(rs.getString("activity_type"))
                        .memberId(rs.getString("member_id"))
                        .memberName(rs.getString("member_name"))
                        .content(rs.getString("content"))
                        .activityTime(rs.getTimestamp("activity_time").toLocalDateTime())
                        .build(),
                businessPlaceId, limit);
    }

    /**
     * 오늘의 예약 일정 조회
     */
    public List<TodayScheduleDTO> getTodaySchedule(String businessPlaceId, Integer limit) {
        if (limit == null || limit <= 0) {
            limit = 10;
        }

        LocalDate today = LocalDate.now();
        List<Reservation> reservations = reservationRepository
                .findByBusinessPlaceIdAndReservationDateOrderByReservationTimeAsc(businessPlaceId, today);

        // PENDING, CONFIRMED 상태만 필터링하고 limit 적용
        final int finalLimit = limit;
        return reservations.stream()
                .filter(r -> r.getStatus() == Reservation.ReservationStatus.PENDING ||
                             r.getStatus() == Reservation.ReservationStatus.CONFIRMED)
                .limit(finalLimit)
                .map(r -> TodayScheduleDTO.builder()
                        .reservationId(r.getId())
                        .memberId(r.getMemberId())
                        .memberName(r.getMember() != null ? r.getMember().getName() : "알 수 없음")
                        .reservationTime(r.getReservationTime())
                        .serviceType(r.getServiceType())
                        .durationMinutes(r.getDurationMinutes())
                        .status(r.getStatus())
                        .notes(r.getNotes())
                        .build())
                .collect(Collectors.toList());
    }

    /**
     * 회원 등록 추이 조회
     */
    public ChartDataDTO.MemberRegistrationTrendDTO getMemberRegistrationTrend(
            String businessPlaceId, Integer days) {
        if (days == null || days <= 0) {
            days = 7;
        }

        LocalDate endDate = LocalDate.now();
        LocalDate startDate = endDate.minusDays(days - 1);

        String sql = "SELECT DATE(created_at) as date, COUNT(*) as count " +
                     "FROM members " +
                     "WHERE business_place_id = ? " +
                     "AND DATE(created_at) BETWEEN ? AND ? " +
                     "GROUP BY DATE(created_at) " +
                     "ORDER BY date";

        List<ChartDataDTO.TimeSeriesDataPoint> dataPoints = jdbcTemplate.query(sql,
                (rs, rowNum) -> ChartDataDTO.TimeSeriesDataPoint.builder()
                        .date(rs.getDate("date").toLocalDate())
                        .count(rs.getInt("count"))
                        .build(),
                businessPlaceId, startDate, endDate);

        // Fill missing dates with 0
        List<ChartDataDTO.TimeSeriesDataPoint> filledDataPoints = new ArrayList<>();
        Map<LocalDate, Integer> dataMap = new HashMap<>();
        for (ChartDataDTO.TimeSeriesDataPoint point : dataPoints) {
            dataMap.put(point.getDate(), point.getCount());
        }

        int totalNewMembers = 0;
        for (LocalDate date = startDate; !date.isAfter(endDate); date = date.plusDays(1)) {
            int count = dataMap.getOrDefault(date, 0);
            totalNewMembers += count;
            filledDataPoints.add(ChartDataDTO.TimeSeriesDataPoint.builder()
                    .date(date)
                    .count(count)
                    .build());
        }

        return ChartDataDTO.MemberRegistrationTrendDTO.builder()
                .dataPoints(filledDataPoints)
                .totalNewMembers(totalNewMembers)
                .build();
    }

    /**
     * 회원 등급별 분포 조회
     */
    public ChartDataDTO.MemberGradeDistributionDTO getMemberGradeDistribution(
            String businessPlaceId) {
        String sql = "SELECT grade, COUNT(*) as count " +
                     "FROM members " +
                     "WHERE business_place_id = ? " +
                     "AND grade IS NOT NULL " +
                     "GROUP BY grade";

        List<Map<String, Object>> results = jdbcTemplate.queryForList(sql, businessPlaceId);

        Map<String, Integer> distribution = new HashMap<>();
        int totalMembers = 0;

        for (Map<String, Object> row : results) {
            String grade = (String) row.get("grade");
            Integer count = ((Number) row.get("count")).intValue();
            distribution.put(grade, count);
            totalMembers += count;
        }

        // Ensure all grades are present
        for (String grade : new String[]{"VIP", "GOLD", "SILVER", "BRONZE", "GENERAL"}) {
            distribution.putIfAbsent(grade, 0);
        }

        return ChartDataDTO.MemberGradeDistributionDTO.builder()
                .distribution(distribution)
                .totalMembers(totalMembers)
                .build();
    }

    /**
     * 예약 추이 조회
     */
    public ChartDataDTO.ReservationTrendDTO getReservationTrend(
            String businessPlaceId, Integer days) {
        if (days == null || days <= 0) {
            days = 7;
        }

        LocalDate endDate = LocalDate.now();
        LocalDate startDate = endDate.minusDays(days - 1);

        String sql = "SELECT reservation_date as date, COUNT(*) as count " +
                     "FROM reservations " +
                     "WHERE business_place_id = ? " +
                     "AND reservation_date BETWEEN ? AND ? " +
                     "GROUP BY reservation_date " +
                     "ORDER BY reservation_date";

        List<ChartDataDTO.TimeSeriesDataPoint> dataPoints = jdbcTemplate.query(sql,
                (rs, rowNum) -> ChartDataDTO.TimeSeriesDataPoint.builder()
                        .date(rs.getDate("date").toLocalDate())
                        .count(rs.getInt("count"))
                        .build(),
                businessPlaceId, startDate, endDate);

        // Fill missing dates with 0
        List<ChartDataDTO.TimeSeriesDataPoint> filledDataPoints = new ArrayList<>();
        Map<LocalDate, Integer> dataMap = new HashMap<>();
        for (ChartDataDTO.TimeSeriesDataPoint point : dataPoints) {
            dataMap.put(point.getDate(), point.getCount());
        }

        int totalReservations = 0;
        for (LocalDate date = startDate; !date.isAfter(endDate); date = date.plusDays(1)) {
            int count = dataMap.getOrDefault(date, 0);
            totalReservations += count;
            filledDataPoints.add(ChartDataDTO.TimeSeriesDataPoint.builder()
                    .date(date)
                    .count(count)
                    .build());
        }

        return ChartDataDTO.ReservationTrendDTO.builder()
                .dataPoints(filledDataPoints)
                .totalReservations(totalReservations)
                .build();
    }

    /**
     * 메모 작성 통계 조회
     */
    public ChartDataDTO.MemoStatisticsDTO getMemoStatistics(
            String businessPlaceId, Integer days) {
        if (days == null || days <= 0) {
            days = 7;
        }

        // 전체 메모 개수
        String totalSql = "SELECT COUNT(*) FROM memos m " +
                          "JOIN members mb ON m.member_id = mb.id " +
                          "WHERE mb.business_place_id = ?";
        Integer totalMemos = jdbcTemplate.queryForObject(totalSql, Integer.class, businessPlaceId);

        // 중요 메모 개수
        String importantSql = "SELECT COUNT(*) FROM memos m " +
                              "JOIN members mb ON m.member_id = mb.id " +
                              "WHERE mb.business_place_id = ? AND m.is_important = true";
        Integer importantMemos = jdbcTemplate.queryForObject(importantSql, Integer.class, businessPlaceId);

        // 아카이브된 메모 개수
        String archivedSql = "SELECT COUNT(*) FROM memos m " +
                             "JOIN members mb ON m.member_id = mb.id " +
                             "WHERE mb.business_place_id = ? AND m.is_archived = true";
        Integer archivedMemos = jdbcTemplate.queryForObject(archivedSql, Integer.class, businessPlaceId);

        // 일별 메모 작성 추이
        LocalDate endDate = LocalDate.now();
        LocalDate startDate = endDate.minusDays(days - 1);

        String dailySql = "SELECT DATE(m.created_at) as date, COUNT(*) as count " +
                          "FROM memos m " +
                          "JOIN members mb ON m.member_id = mb.id " +
                          "WHERE mb.business_place_id = ? " +
                          "AND DATE(m.created_at) BETWEEN ? AND ? " +
                          "GROUP BY DATE(m.created_at) " +
                          "ORDER BY date";

        List<ChartDataDTO.TimeSeriesDataPoint> dataPoints = jdbcTemplate.query(dailySql,
                (rs, rowNum) -> ChartDataDTO.TimeSeriesDataPoint.builder()
                        .date(rs.getDate("date").toLocalDate())
                        .count(rs.getInt("count"))
                        .build(),
                businessPlaceId, startDate, endDate);

        // Fill missing dates with 0
        List<ChartDataDTO.TimeSeriesDataPoint> filledDataPoints = new ArrayList<>();
        Map<LocalDate, Integer> dataMap = new HashMap<>();
        for (ChartDataDTO.TimeSeriesDataPoint point : dataPoints) {
            dataMap.put(point.getDate(), point.getCount());
        }

        for (LocalDate date = startDate; !date.isAfter(endDate); date = date.plusDays(1)) {
            int count = dataMap.getOrDefault(date, 0);
            filledDataPoints.add(ChartDataDTO.TimeSeriesDataPoint.builder()
                    .date(date)
                    .count(count)
                    .build());
        }

        return ChartDataDTO.MemoStatisticsDTO.builder()
                .totalMemos(totalMemos != null ? totalMemos : 0)
                .importantMemos(importantMemos != null ? importantMemos : 0)
                .archivedMemos(archivedMemos != null ? archivedMemos : 0)
                .dailyMemos(filledDataPoints)
                .build();
    }
}

File: voca_crm/lib/data/datasource/statistics_service.dart
import 'dart:convert';

import 'package:voca_crm/core/network/api_client.dart';

class HomeStatistics {
  final String businessPlaceId;
  final String businessPlaceName;
  final int todayReservations;
  final int todayVisits;
  final int pendingMemos;
  final int totalMembers;

  HomeStatistics({
    required this.businessPlaceId,
    required this.businessPlaceName,
    required this.todayReservations,
    required this.todayVisits,
    required this.pendingMemos,
    required this.totalMembers,
  });

  factory HomeStatistics.fromJson(Map<String, dynamic> json) {
    return HomeStatistics(
      businessPlaceId: json['businessPlaceId'] ?? '',
      businessPlaceName: json['businessPlaceName'] ?? '사업장 없음',
      todayReservations: json['todayReservations'] ?? 0,
      todayVisits: json['todayVisits'] ?? 0,
      pendingMemos: json['pendingMemos'] ?? 0,
      totalMembers: json['totalMembers'] ?? 0,
    );
  }
}

class TimeSeriesDataPoint {
  final DateTime date;
  final int count;

  TimeSeriesDataPoint({
    required this.date,
    required this.count,
  });

  factory TimeSeriesDataPoint.fromJson(Map<String, dynamic> json) {
    return TimeSeriesDataPoint(
      date: DateTime.parse(json['date']),
      count: json['count'] ?? 0,
    );
  }
}

class MemberRegistrationTrend {
  final List<TimeSeriesDataPoint> dataPoints;
  final int totalNewMembers;

  MemberRegistrationTrend({
    required this.dataPoints,
    required this.totalNewMembers,
  });

  factory MemberRegistrationTrend.fromJson(Map<String, dynamic> json) {
    return MemberRegistrationTrend(
      dataPoints: (json['dataPoints'] as List)
          .map((item) => TimeSeriesDataPoint.fromJson(item))
          .toList(),
      totalNewMembers: json['totalNewMembers'] ?? 0,
    );
  }
}

class MemberGradeDistribution {
  final Map<String, int> distribution;
  final int totalMembers;

  MemberGradeDistribution({
    required this.distribution,
    required this.totalMembers,
  });

  factory MemberGradeDistribution.fromJson(Map<String, dynamic> json) {
    final Map<String, int> dist = {};
    (json['distribution'] as Map<String, dynamic>).forEach((key, value) {
      dist[key] = value as int;
    });
    return MemberGradeDistribution(
      distribution: dist,
      totalMembers: json['totalMembers'] ?? 0,
    );
  }
}

class ReservationTrend {
  final List<TimeSeriesDataPoint> dataPoints;
  final int totalReservations;

  ReservationTrend({
    required this.dataPoints,
    required this.totalReservations,
  });

  factory ReservationTrend.fromJson(Map<String, dynamic> json) {
    return ReservationTrend(
      dataPoints: (json['dataPoints'] as List)
          .map((item) => TimeSeriesDataPoint.fromJson(item))
          .toList(),
      totalReservations: json['totalReservations'] ?? 0,
    );
  }
}

class MemoStatistics {
  final int totalMemos;
  final int importantMemos;
  final int archivedMemos;
  final List<TimeSeriesDataPoint> dailyMemos;

  MemoStatistics({
    required this.totalMemos,
    required this.importantMemos,
    required this.archivedMemos,
    required this.dailyMemos,
  });

  factory MemoStatistics.fromJson(Map<String, dynamic> json) {
    return MemoStatistics(
      totalMemos: json['totalMemos'] ?? 0,
      importantMemos: json['importantMemos'] ?? 0,
      archivedMemos: json['archivedMemos'] ?? 0,
      dailyMemos: (json['dailyMemos'] as List)
          .map((item) => TimeSeriesDataPoint.fromJson(item))
          .toList(),
    );
  }
}

class RecentActivity {
  final String activityId;
  final String activityType; // MEMO or VISIT
  final String memberId;
  final String memberName;
  final String content;
  final DateTime activityTime;

  RecentActivity({
    required this.activityId,
    required this.activityType,
    required this.memberId,
    required this.memberName,
    required this.content,
    required this.activityTime,
  });

  factory RecentActivity.fromJson(Map<String, dynamic> json) {
    return RecentActivity(
      activityId: json['activityId'] ?? '',
      activityType: json['activityType'] ?? 'MEMO',
      memberId: json['memberId'] ?? '',
      memberName: json['memberName'] ?? '',
      content: json['content'] ?? '',
      activityTime: DateTime.parse(json['activityTime']),
    );
  }
}

class TodaySchedule {
  final String reservationId;
  final String memberId;
  final String memberName;
  final String reservationTime; // HH:mm:ss format
  final String? serviceType;
  final int durationMinutes;
  final String status;
  final String? notes;

  TodaySchedule({
    required this.reservationId,
    required this.memberId,
    required this.memberName,
    required this.reservationTime,
    this.serviceType,
    required this.durationMinutes,
    required this.status,
    this.notes,
  });

  factory TodaySchedule.fromJson(Map<String, dynamic> json) {
    return TodaySchedule(
      reservationId: json['reservationId'] ?? '',
      memberId: json['memberId'] ?? '',
      memberName: json['memberName'] ?? '',
      reservationTime: json['reservationTime'] ?? '',
      serviceType: json['serviceType'],
      durationMinutes: json['durationMinutes'] ?? 60,
      status: json['status'] ?? 'PENDING',
      notes: json['notes'],
    );
  }

  /// 시간 포맷팅 (HH:mm)
  String get formattedTime {
    final parts = reservationTime.split(':');
    if (parts.length >= 2) {
      return '${parts[0]}:${parts[1]}';
    }
    return reservationTime;
  }
}

class StatisticsService {
  final ApiClient _apiClient;

  StatisticsService({ApiClient? apiClient})
      : _apiClient = apiClient ?? ApiClient.instance;

  /// API 응답에서 사용자 친화적 오류 메시지 추출
  String _extractErrorMessage(String responseBody, String fallbackMessage) {
    try {
      final data = jsonDecode(responseBody);
      if (data is Map<String, dynamic>) {
        if (data['fieldErrors'] is Map && (data['fieldErrors'] as Map).isNotEmpty) {
          final fieldErrors = data['fieldErrors'] as Map;
          return fieldErrors.values.first.toString();
        }
        if (data['message'] != null && data['message'].toString().isNotEmpty) {
          return data['message'].toString();
        }
      }
    } catch (_) {}
    return fallbackMessage;
  }

  Future<HomeStatistics> getHomeStatistics(String businessPlaceId) async {
    final response = await _apiClient.get('/api/statistics/home/$businessPlaceId');

    if (response.statusCode == 200) {
      return HomeStatistics.fromJson(jsonDecode(response.body));
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '홈 통계를 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }

  Future<List<RecentActivity>> getRecentActivities(
    String businessPlaceId, {
    int limit = 10,
  }) async {
    final response = await _apiClient.get(
      '/api/statistics/recent-activities/$businessPlaceId',
      queryParams: {'limit': limit.toString()},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((json) => RecentActivity.fromJson(json)).toList();
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '최근 활동을 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }

  /// 오늘의 예약 일정 조회
  Future<List<TodaySchedule>> getTodaySchedule(
    String businessPlaceId, {
    int limit = 10,
  }) async {
    final response = await _apiClient.get(
      '/api/statistics/today-schedule/$businessPlaceId',
      queryParams: {'limit': limit.toString()},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((json) => TodaySchedule.fromJson(json)).toList();
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '오늘의 일정을 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }

  /// 회원 등록 추이 조회
  Future<MemberRegistrationTrend> getMemberRegistrationTrend(
    String businessPlaceId, {
    int days = 7,
  }) async {
    final response = await _apiClient.get(
      '/api/statistics/member-registration-trend/$businessPlaceId',
      queryParams: {'days': days.toString()},
    );

    if (response.statusCode == 200) {
      return MemberRegistrationTrend.fromJson(jsonDecode(response.body));
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '회원 등록 추이를 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }

  /// 회원 등급별 분포 조회
  Future<MemberGradeDistribution> getMemberGradeDistribution(
    String businessPlaceId,
  ) async {
    final response = await _apiClient.get(
      '/api/statistics/member-grade-distribution/$businessPlaceId',
    );

    if (response.statusCode == 200) {
      return MemberGradeDistribution.fromJson(jsonDecode(response.body));
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '회원 등급 분포를 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }

  /// 예약 추이 조회
  Future<ReservationTrend> getReservationTrend(
    String businessPlaceId, {
    int days = 7,
  }) async {
    final response = await _apiClient.get(
      '/api/statistics/reservation-trend/$businessPlaceId',
      queryParams: {'days': days.toString()},
    );

    if (response.statusCode == 200) {
      return ReservationTrend.fromJson(jsonDecode(response.body));
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '예약 추이를 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }

  /// 메모 작성 통계 조회
  Future<MemoStatistics> getMemoStatistics(
    String businessPlaceId, {
    int days = 7,
  }) async {
    final response = await _apiClient.get(
      '/api/statistics/memo-statistics/$businessPlaceId',
      queryParams: {'days': days.toString()},
    );

    if (response.statusCode == 200) {
      return MemoStatistics.fromJson(jsonDecode(response.body));
    } else {
      throw StatisticsServiceException(
        _extractErrorMessage(response.body, '메모 통계를 불러오는 중 오류가 발생했습니다.'),
      );
    }
  }
}

/// 통계 서비스 예외
class StatisticsServiceException implements Exception {
  final String message;

  StatisticsServiceException(this.message);

  @override
  String toString() => message;
}
